### PR TITLE
Close TCP connection when errors prevent handling messages

### LINKED ISSUE #5438: Close TCP connection when errors prevent handling messages
#### Problem

Under `TCPBase::ProcessReceivedBuffer()`, it may not be possible to pass a message to higher layers, e.g. if memory allocation fails or the message is too long. In such cases there is currently no signal to the sender that the message will not be handled.

#### Proposed Solution

Close the TCP connection if the message cannot be handled.

In the case of an over-long message, close as soon as the length is known, rather than waiting to try to allocate buffers for the whole thing.

### LINKED ISSUE #5488: This can be removed once issue #5438 guarantees that the
https://github.com/project-chip/connectedhomeip/blob/3e8a7c7c6d3949c746ffc5e8e50963c0472c376c/src/transport/raw/TCP.h#L134-L144

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in 3e8a7c7c6d3949c746ffc5e8e50963c0472c376c when #5265 was merged. cc @kpschoedel.

### PR DESCRIPTION
**Pull request by kpschoedel**
#### Problem

Under `TCPBase::ProcessReceivedBuffer()`, it may not be possible to
pass a message to higher layers, e.g. if memory allocation fails or the
message is too long. In such cases there is currently no signal to the
sender that the message will not be handled.

#### Summary of Changes

- Close TCP connection in `TCPEndPoint::DriveReceiving` if the
  `OnDataReceived` handler returns an error.
- Introduce `kMaxMessageSize` for the message length limit,
  current based on the maximum packet buffer length (pending #5487).
- Remove `TCPBase` discard state since there will be nothing to discard.

Fixes #5438 - Close TCP connection when errors prevent handling messages
Fixes #5488 - This can be removed once issue #5438 guarantees that the…

### CODE REVIEW

**Comment by andy31415** [comment]
You removed more lines than you added while adding more functionality. I like this patch a lot.

**Review by andy31415 — APPROVED** [review]

File: src/transport/raw/TCP.cpp [hunk 1]
```diff
@@ -309,29 +303,25 @@ CHIP_ERROR TCPBase::ProcessReceivedBuffer(Inet::TCPEndPoint * endPoint, const Pe
             return err;
```
Comment (bzbarsky-apple) [inline src/transport/raw/TCP.cpp:303]: This is returning a CHIP_ERROR as an INET_ERROR.  That doesn't seem right.

And above, where we return CHIP_NO_ERROR we should have INET_NO_ERROR, right?

File: src/transport/raw/TCP.cpp [hunk 2]
```diff
@@ -353,7 +343,7 @@ CHIP_ERROR TCPBase::ProcessSingleMessage(const PeerAddress & peerAddress, Active
         message = System::PacketBufferHandle::New(messageSize, 0);
         if (message.IsNull())
         {
-            return CHIP_ERROR_NO_MEMORY;
+            return INET_ERROR_NO_MEMORY;
         }
         CHIP_ERROR err = state->mReceived->Read(message->Start(), messageSize);
```
Comment (bzbarsky-apple) [inline src/transport/raw/TCP.cpp:348]: This ends up being possibly-returned via `ReturnErrorOnFailure`, but it's not an INET_ERROR.

File: src/transport/raw/TCP.cpp [hunk 3]
```diff
@@ -365,10 +355,10 @@ CHIP_ERROR TCPBase::ProcessSingleMessage(const PeerAddress & peerAddress, Active
     ReturnErrorOnFailure(header.DecodeAndConsume(message));
```
Comment (bzbarsky-apple) [inline src/transport/raw/TCP.cpp:355]: Again, this is potentially reporting a CHIP_ERROR as INET_ERROR.

**Review by LuDuda — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]

File: src/transport/raw/TCP.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply src/transport/raw/TCP.cpp:303]: Yeah, solving this in C++ without making all the error types struct wrappers for integers is a PITA... :(

**Pull request merged**